src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -16,14 +16,235 @@
+
+  /* fork generation count */
+  NaClXMutexLock(&nap_master->children_mu);
+  fork_num++;
+  NaClXMutexUnlock(&nap_master->children_mu);
+
+  /* store cage_ids in both master and parent to provide redundancy and avoid orphans */
+  for (struct NaClApp *nap_cur = nap_master; nap_cur; nap_cur = nap_parent) {
+    NaClXMutexLock(&nap_cur->children_mu);
+    if (nap_cur->children_ids[nap_cur->num_children]) {
+       nap_cur->num_children++;
+    }
+    /* make sure cage_id is unique and assign it to child */
+    InitializeCage(nap_child, nap_master->cage_id + fork_num);
+    NaClLog(1, "[nap %d] incrementing num_children\n", nap_cur->cage_id);
+    nap_cur->children_ids[nap_cur->num_children++] = nap_child->cage_id;
+    if (!DynArraySet(&nap_cur->children, nap_child->cage_id, nap_child)) {
+      NaClLog(LOG_FATAL, "[nap %u] failed to add cage_id %d\n", nap_cur->cage_id, nap_child->cage_id);
+    }
+    NaClLog(1, "[nap %d] new child count: %d\n", nap_cur->cage_id, nap_cur->num_children);
+    NaClXMutexUnlock(&nap_cur->children_mu);
+
+    /* break after parent's cleanup */
+    if (nap_cur == nap_parent) {
+      break;
+    }
+  }
+
+  NaClAppInitialDescriptorHookup(nap_child);
+  NaClLog(1, "fork_num = %d, cage_id = %d\n", fork_num, nap_child->cage_id);
+  if ((*mod_status = NaClAppLoadFileFromFilename(nap_child, nap_child->nacl_file)) != LOAD_OK) {
+    NaClLog(1, "Error while loading \"%s\": %s\n", nap_child->nacl_file, NaClErrorString(*mod_status));
+    NaClLog(LOG_FATAL, "%s\n%s\n",
+                       "Using the wrong type of nexe (nacl-x86-32 on an x86-64 or vice versa) ",
+                       "or a corrupt nexe file may be responsible for this error.");
+  }
+
+  if ((*mod_status = NaClAppPrepareToLaunch(nap_child)) != LOAD_OK) {
+    NaClLog(LOG_FATAL, "Failed to prepare child nap_parent for launch\n");
+  }
+  NaClLog(1, "Loading blob file %s\n", nap_child->nacl_file);
+  if (!nap_child->validator->readonly_text_implemented) {
+    NaClLog(LOG_FATAL, "fixed_feature_cpu_mode is not supported\n");
+  }
+
+  NaClLog(1, "%s\n", "Enabling Fixed-Feature CPU Mode");
+  nap_child->fixed_feature_cpu_mode = 1;
+  if (!nap_child->validator->FixCPUFeatures(nap_child->cpu_features)) {
+    NaClLog(LOG_FATAL, "This CPU lacks features required by fixed-function CPU mode.\n");
+  }
+  if (!NaClAppLaunchServiceThreads(nap_child)) {
+    NaClLog(LOG_FATAL, "Launch service threads failed\n");
+  }
+
+  /* duplicate file descriptor table */
+  for (int old_fd = 0, new_fd; old_fd < CAGING_FD_NUM; old_fd++) {
+    struct NaClDesc *old_nd;
+    old_nd = NaClGetDesc(nap_parent, old_fd);
+    if (!old_nd) {
+      break;
+    }
+    new_fd = NaClSetAvail(nap_child, old_nd);
+    NaClSetDesc(nap_child, new_fd, old_nd);
+    fd_cage_table[nap_child->cage_id][new_fd] = fd_cage_table[nap_parent->cage_id][old_fd];
+    nap_child->fd++;
+    NaClLog(1, "NaClGetDesc() copied parent fd [%d] to child fd [%d]\n", old_fd, new_fd);
```

Comment:
`... [%d] ...` I think introduces yet a new debug message sub-format.

Reply:
Yeah the debug format is a little consistent, I'll see if I can figure out a slightly cleaner way to do this.